core/src/main/java/io/grpc/internal/KeepAliveManager.java

```diff
@@ -170,8 +171,10 @@ public synchronized void onDataReceived() {
       }
       // schedule a new ping
       state = State.PING_SCHEDULED;
-      pingFuture =
-          scheduler.schedule(sendPing, keepAliveTimeInNanos, TimeUnit.NANOSECONDS);
+      if (pingFuture == null) {
```

Comment:
I agree that we should leave this as-is. Mixing `pingFuture` checks with state machine checks makes the code complicated and easily broken (as is evidenced by the change breaking the `IDLE` case).